### PR TITLE
[CTRLRFC-980] Dhyana evolution

### PR DESCRIPTION
**Pull request by goldensgui**
Add Frame rate attribute and implement trigger output

### CODE REVIEW

File: specifics/Dhyana/Dhyana.h [hunk 1]
```diff
@@ -1,291 +1,506 @@
...
+/*----- PROTECTED REGION ID(Dhyana::Additional Class Declarations) ENABLED START -----*/
+
+enum class TriggeroutMode : short
+{
+	EXPOSURESTART,
+	READOUT,
+	GLOBAL
+};
+
+enum class edgeMode : short
```
Comment (soleil-ica) [inline specifics/Dhyana/Dhyana.h:124]: Set the 1st letter in CAPITAL case

File: specifics/Dhyana/Dhyana.h [hunk 2]
```diff
@@ -1,291 +1,506 @@
...
+} ;
+typedef _edge3Enum edge3Enum;
+
+/*----- PROTECTED REGION ID(Dhyana::Additional Class Declarations) ENABLED START -----*/
+
+enum class TriggeroutMode : short
+{
+	EXPOSURESTART,
+	READOUT,
+	GLOBAL
```
Comment (soleil-ica) [inline specifics/Dhyana/Dhyana.h:121]: add a k and use the mixed case notation, eg: kExposureStart 

idem for all Enums

File: specifics/Dhyana/Dhyana.h [hunk 3]
```diff
@@ -1,291 +1,506 @@
...
+	double 					  m_delay_ch3;
+	double 					  m_width_ch1;
+	double 					  m_width_ch2;
+	double 					  m_width_ch3;
+	short 					  m_edge_ch1;
+	short 					  m_edge_ch2;
+	short 					  m_edge_ch3;
+	lima::Dhyana::Camera::TucamSignal signal1;
+	lima::Dhyana::Camera::TucamSignal signal2;
+	lima::Dhyana::Camera::TucamSignal signal3;
```
Comment (soleil-ica) [inline specifics/Dhyana/Dhyana.h:494]: add a m_ before the members names

File: specifics/Dhyana/DhyanaStateMachine.cpp [hunk 4]
```diff
@@ -1,212 +1,472 @@
...
-		if ( get_state()==Tango::FAULT && is_device_initialized() )
-		{
-           return true;
-		}		
-		//	Re-Start of Generated Code
-		return false;
+		/*----- PROTECTED REGION END -----*/	//	Dhyana::tucamVersionStateAllowed_READ
+			return false;
+		}
+		return true;
```
Comment (soleil-ica) [inline specifics/Dhyana/DhyanaStateMachine.cpp:79]: The state machine is wrong (because of Pogo 9): please see with Arafat to manage it together

**Review by soleil-ica — CHANGES_REQUESTED** [review]

File: specifics/Dhyana/Dhyana.cpp [hunk 5]
```diff
@@ -1,720 +1,2126 @@
...
+
+	/*----- PROTECTED REGION ID(Dhyana::get_device_property_after) ENABLED START -----*/
+	
+	//	Check device property data members init
+	PropertyHelper::create_property_if_empty(this, dev_prop,"0","MemorizedTemperatureTarget");	
+	PropertyHelper::create_property_if_empty(this, dev_prop,"1","MemorizedFanSpeed");	
+	PropertyHelper::create_property_if_empty(this, dev_prop,"LOW","MemorizedGlobalGain");	
+	PropertyHelper::create_property_if_empty(this, dev_prop,"1","__ExpertTimerPeriod");
+
+	/*----- PROTECTED REGION END -----*/	//	Dhyana::get_device_property_after
```
[the remark this answers is not in the document]
Comment (soleil-ica) [reply specifics/Dhyana/Dhyana.cpp:452]: Checked with Kewin: they have to be memorised

**Review by soleil-ica — CHANGES_REQUESTED** [review]

File: specifics/Dhyana/DhyanaClass.cpp [hunk 6]
```diff
@@ -812,7 +812,7 @@ void DhyanaClass::attribute_factory(vector<Tango::Attr *> &att_list)
 	
 	delay1->set_default_properties(delay1_prop);
 	//	Not Polled
-	delay1->set_disp_level(Tango::OPERATOR);
+	delay1->set_disp_level(Tango::EXPERT);
```
[the remark this answers is not in the document]
Comment (goldensgui) [reply specifics/Dhyana/DhyanaClass.cpp:815]: Yes, line 800 : delay1_prop.set_unit("ms");

**Review by soleil-ica — APPROVED** [review]

**Review by soleil-ica — APPROVED** [review]

**Pull request merged**